### DISCUSSION TITLE
Unused Imports

### DISCUSSION THREAD

**Discussion by blkmlk** [rust-lang/rustfmt#5722] [Q&A] [answered]
Hi! Is there a configuration parameter that removes unused imports?

**Comment by wellkilo** [comment, marked as answer]
No. rustfmt is intentionally a formatter, not an unused-code analyzer, so it can reorder or regroup imports but does not decide that an import is semantically unused and delete it.

Use the compiler/Clippy fix path instead:

```bash
cargo fix --allow-dirty --allow-staged
# or for the common broader pass
cargo clippy --fix --allow-dirty --allow-staged --all-targets --all-features
```

Review the diff afterwards, especially when conditional compilation or feature-specific imports are involved. IDE “organize imports” actions are another option, but they are separate from rustfmt.

The related rustfmt settings only control formatting:

- `reorder_imports = true|false` alphabetizes import groups.
- `imports_granularity` controls how imports are merged/grouped.

Neither performs unused-import analysis. Current configuration reference: [`reorder_imports`](https://github.com/rust-lang/rustfmt/blob/9e92f7b48cf49412c647eb6b20b6e4315481a23f/Configurations.md#reorder_imports) and [`imports_granularity`](https://github.com/rust-lang/rustfmt/blob/9e92f7b48cf49412c647eb6b20b6e4315481a23f/Configurations.md#imports_granularity).

If this clarifies the formatter/linter boundary, please mark the reply as accepted.
